### ISSUE TITLE
raise code coverage to 100%

### ISSUE THREAD

**Issue by reactive-firewall** [reactive-firewall/multicast#53]
# Coverage is below 100% still.

The goal of this issue is to raise the overall test coverage of code in the project's multicast module (e.g. code located in "multicast/*.py") to above 99% (basically 100% when rounding)

Work on this goal is to be tracked by this GH issue.

Notes:
 * Development branch: testing-coverage
 * Tests in the CI job "COVERAGE" and "INTEGRATION" report their coverage to the service codecov.
 * Testing Naming guidelines are detailed in [CEP-9](https://gist.github.com/reactive-firewall/d840ee9990e65f302ce2a8d78ebe73f6)
 * Test code should be documented as per [CEP-7](https://gist.github.com/reactive-firewall/123b8a45f1bdeb064079e0524a29ec20)

**Comment by reactive-firewall** [comment]
@coderabbitai, let's bring issue #53 up-to-date with recent changes:
1. Please look at current test code coverage for the project and point out any coverage gaps.
2. Consider how coverage might be improved

**Comment by reactive-firewall** [comment]
@coderabbitai, please elaborate more about possible specific test implementations for testing both `multicast/hear.py` and `multicast/recv.py` with a focus on currently untested portions of the code.

**Comment by reactive-firewall** [comment]
@coderabbtiai, Sorry for the confusion, but consider, by "hear" and "recv", I meant all of the code in `multicast/hear.py` and `multicast/recv.py` from this project. Please read all of the python files in the project, and please read the suplementry documentation in `README.md` and `docs/USAGE.md`, then please try again to:

> please elaborate more about possible specific test implementations for testing both `multicast/hear.py` and `multicast/recv.py` with a focus on currently untested portions of the code.

Notes:
* current test code can be found in tests/
* you can find supplemental testing in the various doctests throughout the codebase.
* consider edge-case testing
* consider mocking could be done for multicast by mocking how the actual sockets are generated in multicast.skt (from `multicast/skt.py`), and implementing mocking is within the scope of this issue, and explore how to implement it as part of new testing.
* Consider this is a complex problem and consider using a backtracking strategy, and dynamic programing when planning what project code needs new test-cases the most.
* Consider the primary goal is to ensure 100% of python code is covered by test code
* assume the following report regarding the current coverage of the default (ie master) branch:
```plain
Name                    Stmts   Miss Branch BrPart  Cover   Missing
-------------------------------------------------------------------
multicast/__init__.py      63      0     10      0   100%
multicast/__main__.py      83      1     22      1    98%   482
multicast/hear.py          75     36     16      0    47%   246-248, 265-269, 280-282, 297-299, 315-332, 408-429, 507-511
multicast/recv.py          75     19     14      3    73%   277, 281, 285-290, 346-349, 493-515
multicast/send.py          40      0      8      0   100%
multicast/skt.py           18      0      2      0   100%
-------------------------------------------------------------------
TOTAL                     354     56     72      4    83%
```

**Comment by reactive-firewall** [comment]
@coderabbitai, Most of these tests failed for one reason or another, I've started debugging them, and am on the test: 'tests/test_hear_missing_params.py' which is logicly wrong; if group and groups are omitted they do NOT throw an error, instead they use the default value from `multicast._MCAST_DEFAULT_GROUP` which in the case of `multicast.hear.McastHEAR` instances do just that when missing the group keyword argument; this hangs the test because no ValueError is ever raised. Worse yet the `multicast.hear.McastHEAR` instance is never properly stopped in this test.

Let's review which lines we're testing:
- Covers lines 246-248 where exceptions are raised if mandatory parameters are missing.
- those lines are the class method [McastServer.server_activate‎](https://github.com/reactive-firewall/multicast/blob/71c59bac7325154f0d02f8d1e5ff1ebe66c0850a/multicast/hear.py#L246-L248)
- Let's test those more directly: create a McastServer instance and call the "server_activate‎" to test the overloaded implementation where we can validate the call eventually reached the superclass `socketserver.UDPServer`'s `server_activate‎` function simply by checking the absence of errors when using the constructor, unless I'm mistaken about what the documentation says. However we can sidestep a lot of the doubt and focus on the call to open_for_request which will in turn alter the instance's `socket` attribute.
So to summarize we can create an instance and via the constructor validate the socket attribute is not None, and that the socket attribute is further of the UDP protocol and type. Then we can clean up the server by shutting it down before entering daemon mode.

double check my logic and then proceded to re-design the entire testcase around the goal "test the effect of overloading server_activate to cover lines 246-248 in multicast/hear.py" and use my new plan as a starting point to implement the replacement test code.

**Comment by reactive-firewall** [comment]
@coderabbitai, this is much better, but do we actually need the mocking if our goal is to cover the actual lines of code? Also please consider [CEP-7](https://gist.github.com/reactive-firewall/123b8a45f1bdeb064079e0524a29ec20) [CEP-8](https://gist.github.com/reactive-firewall/b7ee98df9e636a51806e62ef9c4ab161) [CEP-9](https://gist.github.com/reactive-firewall/d840ee9990e65f302ce2a8d78ebe73f6) while generating these test cases

**Comment by reactive-firewall** [comment]
@coderabbitai, I have just merged several PR/MR into the master branch, please consider this the start of the next iteration of this issue.

Consider the latest test-coverage results (from [66e7ff4](https://github.com/reactive-firewall/multicast/commit/66e7ff489b75562a02f41f8c09d6fc2a51d9aa83)) to be:

```plain
Name                      Stmts   Miss Branch BrPart  Cover   Missing
---------------------------------------------------------------------
multicast/__init__.py        67      0      4      0   100%
multicast/__main__.py        83      1     10      1    98%   482
multicast/exceptions.py      17      4      0      0    76%   220-226
multicast/hear.py            82      4     12      4    89%   337-338, 342, 420, 424->exit, 440->exit
multicast/recv.py            76     15     10      0    80%   346-349, 493-515
multicast/send.py            56      1     10      1    97%   364
multicast/skt.py             19      0      2      0   100%
---------------------------------------------------------------------
TOTAL                       400     25     48      6    92%
```

1. Please look at the coverage gaps line by line and analyse the relevant code in preparation of designing test cases.
2. for each of the relevant test coverage gaps design a possible strategy for testing that will exercise the code path. Consider multiple strategies when appropriate.
3. Use the strategies to generate possible implementations for each case. Generate both unittest based, and docstring based implementations for each case.
4. Finally Organize and explain each strategy and related implementations.

Extra Considerations:
* consider the coverage configuration in ".coveragerc"
* consider the existing test code and its use of "tests/context.py"
* consider the project's conventions, and overall style while generating implementations.

**Comment by reactive-firewall** [comment]
# WIP

> [!WARNING]
> This comment is in progress

* Consider [multicast/recv.py around line 493](https://app.codecov.io/gh/reactive-firewall/multicast/commit/048089d84551d31e9b16ac254d66678db3314550/blob/multicast/recv.py#L493) of 048089d

  The branch is not yet tested when the given input (parser) is `None`

* Consider [multicast/hear.py around line 326](https://app.codecov.io/gh/reactive-firewall/multicast/commit/048089d84551d31e9b16ac254d66678db3314550/blob/multicast/hear.py#L326) of 048089d

  The branch is not fully tested, especially when the condition is false. Also cause for line 342.

* Consider [multicast/hear.py around line 342](https://app.codecov.io/gh/reactive-firewall/multicast/commit/048089d84551d31e9b16ac254d66678db3314550/blob/multicast/hear.py#L342) of 048089d

  The else of the branch starting around [line 326](https://app.codecov.io/gh/reactive-firewall/multicast/commit/048089d84551d31e9b16ac254d66678db3314550/blob/multicast/hear.py#L326) is not tested and thus [line 342](https://app.codecov.io/gh/reactive-firewall/multicast/commit/048089d84551d31e9b16ac254d66678db3314550/blob/multicast/hear.py#L342) is not tested.

* Consider [multicast/hear.py around line 337](https://app.codecov.io/gh/reactive-firewall/multicast/commit/048089d84551d31e9b16ac254d66678db3314550/blob/multicast/hear.py#L337) and [around line 338](https://app.codecov.io/gh/reactive-firewall/multicast/commit/048089d84551d31e9b16ac254d66678db3314550/blob/multicast/hear.py#L338) of 048089d

  These lines are completely untested. The inner function could be tested by mocking a server-like object that has a function for shutdown_server, and then just pass it to the function as a test.

* Consider [multicast/hear.py around line 424](https://app.codecov.io/gh/reactive-firewall/multicast/commit/048089d84551d31e9b16ac254d66678db3314550/blob/multicast/hear.py#L424) of 048089d

  This line is not tested because data equal to `None` is never passed to the handler with a valid socket during tests. Also line 419 prevents the possibility. This code should probably be refactored.


- WIP -

**Comment by reactive-firewall** [comment]
# Update for new Development cycle

## Report

Current coverage report of `multicast` codebase by tests:

### Report Data

```plain
Name                      Stmts   Miss Branch BrPart  Cover   Missing
---------------------------------------------------------------------
multicast/__init__.py        67      0      4      0   100%
multicast/__main__.py        89      0     12      0   100%
multicast/exceptions.py      26      0      4      0   100%
multicast/hear.py            89      3     12      3    92%   337-338, 342, 424->exit, 440->exit
multicast/recv.py            63      0      8      0   100%
multicast/send.py            60      0     10      0   100%
multicast/skt.py             19      0      2      0   100%
---------------------------------------------------------------------
TOTAL                       413      3     52      3    98%
```

### Codebase mappings

The following are the links to the missing test-coverage lines of code in GitHub for this report data.

 * [multicast/hear.py lines 337-338](https://github.com/reactive-firewall/multicast/blob/v2.0.0-rc-5/multicast/hear.py#L327-L338)
 * [multicast/hear.py lines 326-342](https://github.com/reactive-firewall/multicast/blob/v2.0.0-rc-5/multicast/hear.py#L326-L342) Namely [multicast/hear.py lines 342](https://github.com/reactive-firewall/multicast/blob/v2.0.0-rc-5/multicast/hear.py#L342)
 * [multicast/hear.py lines 373-424](https://github.com/reactive-firewall/multicast/blob/v2.0.0-rc-5/multicast/hear.py#L373-L424) Namely [multicast/hear.py lines 424](https://github.com/reactive-firewall/multicast/blob/v2.0.0-rc-5/multicast/hear.py#L424)
 * [multicast/hear.py lines 424-440](https://github.com/reactive-firewall/multicast/blob/v2.0.0-rc-5/multicast/hear.py#L424-L440) Namely [multicast/hear.py lines 440](https://github.com/reactive-firewall/multicast/blob/v2.0.0-rc-5/multicast/hear.py#L440)

---

## Analysis

### Analysis of coverage gaps in multicast codebase.

1. Consider [multicast/hear.py around line 326](https://app.codecov.io/gh/reactive-firewall/multicast/commit/19f462483ab0ab1e1a3670aafb019f471cfd4500/blob/multicast/hear.py#L326) of git commit 19f462483ab0ab1e1a3670aafb019f471cfd4500

```python
		if request is not None and request[0] is not None and """STOP""" in str(request[0]):
```

  * The conditional branch is not yet tested when the given input (e.g. `request`) is `None`
  * The conditional branch is not yet tested when the given input (e.g. `request[0]`) is not `None` but the value (e.g. `str(request[0])`) does not contain "STOP"
  * This can be directly tested against the `handle_error` method.

2. Consider [multicast/hear.py around line 337](https://app.codecov.io/gh/reactive-firewall/multicast/commit/19f462483ab0ab1e1a3670aafb019f471cfd4500/blob/multicast/hear.py#L337) of git commit 19f462483ab0ab1e1a3670aafb019f471cfd4500

```python
				if a_server is not None:
```

  * The conditional branch is not yet tested when the given input (e.g. `a_server`) is `None`
  * The conditional branch is not yet tested when the given input (e.g. `a_server`) is a class instance that has the method `shutdown(self)`
  * this conditional directly impacts the following coverage gap in [multicast/hear.py around line 338](https://app.codecov.io/gh/reactive-firewall/multicast/commit/19f462483ab0ab1e1a3670aafb019f471cfd4500/blob/multicast/hear.py#L338)

3. Consider [multicast/hear.py around line 338](https://app.codecov.io/gh/reactive-firewall/multicast/commit/19f462483ab0ab1e1a3670aafb019f471cfd4500/blob/multicast/hear.py#L338) of git commit 19f462483ab0ab1e1a3670aafb019f471cfd4500

```python
					a_server.shutdown()
```

  * The trivial call to the input's `shutdown` method can be tested with the above conditional line 337.
  * This line is not yet tested when the given input (e.g. `a_server`) is a class instance that has the method `shutdown(self)`

3. Consider [multicast/hear.py around line 342](https://app.codecov.io/gh/reactive-firewall/multicast/commit/19f462483ab0ab1e1a3670aafb019f471cfd4500/blob/multicast/hear.py#L342) of git commit 19f462483ab0ab1e1a3670aafb019f471cfd4500

```python
			super(McastServer, self).handle_error(request, client_address)
```

  * The trivial call to the input's super `handle_error` method can be tested with the above conditional line 326.
  * This line is not yet tested when the given input (e.g. `request`) is `None`
  * This line is not yet tested when the given input (e.g. `request[0]`) is not `None` but the value (e.g. `str(request[0])`) does not contain "STOP"
  * This can be directly tested against the `handle_error` method too.

4. Consider [multicast/hear.py around line 424](https://app.codecov.io/gh/reactive-firewall/multicast/commit/19f462483ab0ab1e1a3670aafb019f471cfd4500/blob/multicast/hear.py#L424) of git commit 19f462483ab0ab1e1a3670aafb019f471cfd4500

```python
		if data is not None:
```

  * this conditional is never reached in tests because either data or sock fail on line 419 or data is not `None` or coverage is not run on the process.
  * this can be covered by extending the docstring's doc tests with something like:

```python
"""
			Testcase 3: `handle` requires valid requests or ignores input.

			>>> handler.request = ("The Test", multicast.genSocket())
			>>> handler.client_address = ("224.0.1.2", 51234)
			>>> handler.handle() is None
			True
			>>>
"""
```

  * this approach also should cover the missed branch around line 440 at the same time.

---

## AI Tasks

@coderabbitai, please assist with the following tasks:

1. Please read every python file in the recent version of the codebase, tagged as "v2.0.0-rc-5".
2. Please generate possible specific test implementations improve the project's test coverage of the above reported missing lines. Be thorough.

### Extra Considerations

 * consider the coverage configuration in ".coveragerc"
 * consider the existing test code and its use of "tests/context.py"
 * consider the project's conventions, and overall style while generating implementations.
 * unittest test-suites sub-classes should subclass from `tests.context.BasicUsageTestSuite` instead of `unittest.TestCase` as is consistent for test code in the project.
